x-pack/plugins/alerting/server/task_runner/execution_handler.ts

```diff
@@ -0,0 +1,419 @@
+  >;
+  private ruleTypeActionGroups;
+  private taskRunnerContext: TaskRunnerContext;
+  private taskInstance: RuleTaskInstance;
+  private ruleRunMetricsStore: RuleRunMetricsStore;
+  private apiKey: RawRule['apiKey'];
+  private ruleConsumer: string;
+  private executionId: string;
+  private ruleLabel: string;
+  private ephemeralActionsToSchedule: number;
+  private mutedAlertIdsSet: Set<string>;
+  private CHUNK_SIZE = 1000;
+  private skippedAlerts: { [key: string]: { reason: string } } = {};
+
+  constructor({
+    rule,
+    ruleType,
+    logger,
+    alertingEventLogger,
+    taskRunnerContext,
+    taskInstance,
+    ruleRunMetricsStore,
+    apiKey,
+    ruleConsumer,
+    executionId,
+    ruleLabel,
+    request,
+  }: ExecutionHandlerOptions<
+    Params,
+    ExtractedParams,
+    RuleState,
+    State,
+    Context,
+    ActionGroupIds,
+    RecoveryActionGroupId
+  >) {
+    this.logger = logger;
+    this.alertingEventLogger = alertingEventLogger;
+    this.rule = rule;
+    this.ruleType = ruleType;
+    this.ruleTypeActionGroups = new Map(
+      ruleType.actionGroups.map((actionGroup) => [actionGroup.id, actionGroup.name])
+    );
+    this.taskRunnerContext = taskRunnerContext;
+    this.taskInstance = taskInstance;
+    this.ruleRunMetricsStore = ruleRunMetricsStore;
+    this.apiKey = apiKey;
+    this.ruleConsumer = ruleConsumer;
+    this.executionId = executionId;
+    this.ruleLabel = ruleLabel;
+    this.request = request;
+    this.ephemeralActionsToSchedule = taskRunnerContext.maxEphemeralActionsPerRule;
+    this.mutedAlertIdsSet = new Set(rule.mutedInstanceIds);
+  }
+
+  public async run(
+    alerts: Record<string, Alert<State, Context, ActionGroupIds | RecoveryActionGroupId>>,
+    recovered: boolean = false
+  ) {
+    const {
+      CHUNK_SIZE,
+      logger,
+      alertingEventLogger,
+      ruleRunMetricsStore,
+      taskRunnerContext: { actionsConfigMap, actionsPlugin },
+      taskInstance: {
+        params: { spaceId, alertId: ruleId },
+      },
+    } = this;
+
+    const executables = this.generateExecutables({ alerts, actions: this.rule.actions, recovered });
+
+    if (!!executables.length) {
+      const logActions = [];
+      const bulkActions: EnqueueExecutionOptions[] = [];
+      const actionsClient = await this.taskRunnerContext.actionsPlugin.getActionsClientWithRequest(
+        this.request
+      );
+
+      this.ruleRunMetricsStore.incrementNumberOfGeneratedActions(executables.length);
+
+      for (const { action, alert, alertId } of executables) {
+        const { actionTypeId } = action;
+        const actionGroup: ActionGroupIds | RecoveryActionGroupId = recovered
+          ? this.ruleType.recoveryActionGroup.id
+          : alert.getScheduledActionOptions()?.actionGroup!;
+
+        const instanceState = recovered ? {} : alert.getScheduledActionOptions()?.state!;
+
+        if (!recovered) {
+          alert.updateLastScheduledActions(action.group as ActionGroupIds);
```

Comment:
I am not sure about this part.
We were running this function for an alert once, now it would be as much as number of the actions.

Reply:
I think it's fine that it might run multiple times. The first call updates the lastScheduledAction date to `now` so calling it multiple times will just set the date a little later. The second call "unschedules" the actions because the requested action has been scheduled, so calling it multiple times will just set the same variable to `undefined` multiple times. It could be more efficient but I don't think it affects the functionality.